File: Data Structures & Algorithms/binary-search/submission-0.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int left = 0;
        int right = (int)nums.size() - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;  // avoids overflow

            if (nums[mid] == target) {
                return mid;
            } 
            else if (nums[mid] < target) {
                left = mid + 1;   // search right half
            } 
            else {
                right = mid - 1;  // search left half
            }
        }

        return -1;
    }
};

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-0.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        left, right = 0, len(nums) - 1
    
        while left < right:
            mid = left + (right - left) // 2

        # If mid element is greater than the right element, 
        # the minimum must be in the right half.
            if nums[mid] > nums[right]:
                left = mid + 1
        # Otherwise, the minimum is at mid or in the left half.
            else:
                right = mid
            
        return nums[left]

        
        